structure_tests/drivers/docker_driver.go

```diff
@@ -0,0 +1,341 @@
+}
+
+// copies a file from a docker container to the local fs, and returns its path
+// caller is responsible for removing this file when finished
+func (d *DockerDriver) retrieveFile(t *testing.T, path string, directory bool) (string, error) {
+	ctx := context.Background()
+
+	// TODO(nkubala): this contains a hack to get around the fact that
+	// the client will not allow creating a container without a command.
+	// we should remove this, or at the very least find a better alternative
+	// given that not every container is guaranteed to have a shell.
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        d.currentImage,
+			Cmd:          []string{"NOOP_COMMAND_DO_NOT_RUN"},
+			AttachStdout: true,
+			AttachStderr: true,
+		},
+		HostConfig:       nil,
+		NetworkingConfig: nil,
+		Context:          ctx,
+	})
+	if err != nil {
+		t.Errorf("Error creating container: %s", err.Error())
+		return "", err
+	}
+
+	if directory {
+		return "", nil
+	} else {
+		tmpFile, err := ioutil.TempFile("", "structure_test")
+		if err != nil {
+			t.Errorf("Error when creating temp file: %s", err.Error())
+			return "", err
+		}
+		stream := bufio.NewWriter(tmpFile)
+
+		err = d.cli.DownloadFromContainer(container.ID, docker.DownloadFromContainerOptions{
+			OutputStream: stream,
+			Path:         path,
+			Context:      ctx,
+		})
+		if err != nil {
+			t.Errorf("Error when downloading file from container: %s", err.Error())
+			return "", err
+		}
+		stream.Flush()
+		tmpFile.Close()
+		return tmpFile.Name(), nil
+	}
+}
+
+func (d *DockerDriver) StatFile(t *testing.T, path string) (os.FileInfo, error) {
+	file, err := d.retrieveFile(t, path, false)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(file)
+
+	f, err := os.Stat(file)
+	if err != nil {
+		return nil, err
+	}
+	if f == nil {
+		return nil, nil
+	}
+	return f, nil
+}
+
+func (d *DockerDriver) ReadFile(t *testing.T, path string) ([]byte, error) {
+	file, err := d.retrieveFile(t, path, false)
+	if err != nil {
+		return nil, err
+	}
+	defer os.Remove(file)
+	return ioutil.ReadFile(file)
+}
+
+func (d *DockerDriver) ReadDir(t *testing.T, path string) ([]os.FileInfo, error) {
+	// TODO(nkubala): unimplemented
+	return nil, nil
+}
+
+// This method takes a command (in the form of a list of args), and does the following:
+// 1) creates a container, based on the "current latest" image, with the command set as
+// the command to run when the container starts
+// 2) starts the container
+// 3) commits the container with its changes to a new image,
+// and sets that image as the new "current image"
+func (d *DockerDriver) runAndCommit(t *testing.T, env []string, command []string) string {
+
+	// this is a placeholder command since apparently the client doesnt allow creating
+	// a container without a command.
+	// TODO(nkubala): figure out how to remove this
+	if len(command) == 0 {
+		command = []string{"/bin/sh"}
```

Comment:
Could you make this a token placeholder for now? It's fine that it's unused, but naming it something like "UNUSED COMMAND" might make more sense.

Reply:
ah yep meant to do that, thanks